### ISSUE TITLE
Accessing application state within `canActivate` not optimal

### ISSUE THREAD

**Issue by analog-nico** [vuejs/vue-router#533]
I boot my app using `vue-router` and `vuex` – nothing special here:

``` js
let Vue = require('vue'), VueRouter = require('vue-router'), Vuex = require('vuex')

Vue.use(VueRouter)
Vue.use(Vuex)

let App = Vue.extend({
    store: new Vuex.Store({ state: { ... }, mutations: { ... } }),
    vuex: { actions: { ... }, getters: { ... } }
})

let router = new VueRouter()
router.map({ ... })

router.start(App, '#app')
```

In a Vue component – now it is getting interesting – I decide whether to activate it or to do a redirect based on the Vuex-based application state:

``` js
Vue.component('sign-up-form', {
    // ...
    route: {
        canActivate(transition) {

            let signupState = transition.to.router.app.signupState // <-- Vuex getter

            if (signupState === 'start') {
                return true
            } else {
                transition.redirect({ name: 'onboarding' })
            }

        }
    }
})
```

Accessing the application state is actually difficult for `canActivate` because the Vue instance does not exist yet and `this` is `null`. So using `this.signupState` is not an option.

Accessing the application state through the root Vue instance (`app`) is IMO the best way to go. However, using the train wreck `transition.to.router.app` feels not optimal.

Admittedly, using `transition.to.router.app` is no hack: [`transition.to`](http://router.vuejs.org/en/pipeline/hooks.html#transition-object), [`to.router`](http://router.vuejs.org/en/route.html#route-object--route-matching), and [`router.app`](http://router.vuejs.org/en/api/properties.html) are all officially documented.

IMHO `canActivate` will likely produce many cases where the decision can't be properly made since it is not bound to a Vue instance. Therefore, I suggest to give it more context. I would not bind it to `app` since that may easily produce confusion with the other hooks that are bound to the Vue instance being deactivated/reused/activated. Instead, I suggest to add an additional parameter:

``` js
/**
 * @param app – The root Vue instance managed by the router instance.
 */
function canActivate(transition, app) { ... }
```

**Comment by fnlctrl** [comment]
I think it's better just import the store and directly use its state inside transition hooks.
```javascript
import store from '...'
.......

canActivate(transition) {
    let signupState = store.state.signupState // <-- 
}
```

**Comment by analog-nico** [comment]
I thought of this approach, too. However, this would mean I sometimes use the `getters` and sometimes access the `state` directly. And by doing that I instantly **loose** the improved **maintainability** the `getters` introduce.

The [Vuex documentation](http://vuex.vuejs.org/en/tutorial.html#step-4-get-the-value-into-the-component) explains it nicely:

> You might be wondering - why did we choose to use a getter instead of directly accessing the value from the state. [This] presents several distinct advantages:
>
> 1. [...]
> 2. [...]
> 3. If the value is moved from say `store.count` to `store.counter.value` you'd have to update one getter instead of dozens of components.

Extracting the `Vuex.Store` into `store.js` would make sense, though, if the `Vuex.Store` would not only manage the `state` and the `mutations` but also the `actions` and `getters`. If that were the case I would choose this option. I'll open an issue for that...

Until accessing the `getters` through the `Vuex.Store` becomes possible, accessing the `getters` through the `app` is a perfectly fine solution. Since the `router` depends on `app` anyway it doesn't produce much tighter coupling and the maintainability is preserved.

Does anyone see other use cases where exposing `app` as the second parameter of `canActivate` comes in handy? If so, we were more motivated to push this feature request, I guess.

**Comment by fnlctrl** [comment]
You're right, getters have better maintainability.
As a matter of consistency, it would be a little weird if `canActivate` has `app` as parameter while others (`activate`,`data`,etc) do not (they have access to `this` though).

I think the real problem is that vuex getters are dependent on vue instances.
Wouldn't it be better, if vuex store can be standalone (having its own getters), and we can directly access states via its getters, without the presence of vue instances. @yyx990803

**Comment by yyx990803** [comment]
This seems to be resolved in Vuex 2.0

**Comment by analog-nico** [comment]
Indeed, it is. Thanks!
